compiler/AST/LoopExpr.cpp

```diff
@@ -0,0 +1,648 @@
+                                           new SymExpr(sifnIterator),
+                                           new BlockStmt(stmt),
+                                           false, // is it a coforall?
+                                           zippered));
+
+  return sifn;
+}
+
+static FnSymbol* buildLeaderIteratorFn(FnSymbol* fn,
+                                       const char* iteratorName,
+                                       bool zippered)
+{
+  FnSymbol* lifn = new FnSymbol(iteratorName);
+  lifn->addFlag(FLAG_FN_RETURNS_ITERATOR);
+
+  Expr* tag = new SymExpr(gLeaderTag);
+  ArgSymbol* lifnTag = new ArgSymbol(INTENT_PARAM, "tag", dtUnknown,
+                                     new CallExpr(PRIM_TYPEOF, tag));
+  lifn->insertFormalAtTail(lifnTag);
+
+  ArgSymbol* lifnIterator = new ArgSymbol(INTENT_BLANK, "iterator", dtAny);
+  lifn->insertFormalAtTail(lifnIterator);
+
+  lifn->where = new BlockStmt(new CallExpr("==", lifnTag, tag->copy()));
+
+  VarSymbol* leaderIterator = newTempConst("_leaderIterator");
+  leaderIterator->addFlag(FLAG_EXPR_TEMP);
+  lifn->insertAtTail(new DefExpr(leaderIterator));
+
+  if( !zippered ) {
+    lifn->insertAtTail(new CallExpr(PRIM_MOVE, leaderIterator, new CallExpr("_toLeader", lifnIterator)));
+  } else {
+    lifn->insertAtTail(new CallExpr(PRIM_MOVE, leaderIterator, new CallExpr("_toLeaderZip", lifnIterator)));
+  }
+
+  lifn->insertAtTail(new CallExpr(PRIM_RETURN, leaderIterator));
+
+  return lifn;
+}
+
+static FnSymbol* buildFollowerIteratorFn(FnSymbol* fn,
+                                         const char* iteratorName,
+                                         bool zippered,
+                                         VarSymbol*& followerIterator)
+{
+  FnSymbol* fifn = new FnSymbol(iteratorName);
+  fifn->addFlag(FLAG_ITERATOR_FN);
+
+  Expr* tag = new SymExpr(gFollowerTag);
+  ArgSymbol* fifnTag = new ArgSymbol(INTENT_PARAM, "tag", dtUnknown,
+                                     new CallExpr(PRIM_TYPEOF, tag));
+  fifn->insertFormalAtTail(fifnTag);
+
+  ArgSymbol* fifnFollower = new ArgSymbol(INTENT_BLANK, iterFollowthisArgname, dtAny);
+  fifn->insertFormalAtTail(fifnFollower);
+
+  ArgSymbol* fifnIterator = new ArgSymbol(INTENT_BLANK, "iterator", dtAny);
+  fifn->insertFormalAtTail(fifnIterator);
+
+  fifn->where = new BlockStmt(new CallExpr("==", fifnTag, tag->copy()));
+  followerIterator = newTempConst("_followerIterator");
+  followerIterator->addFlag(FLAG_EXPR_TEMP);
+  fifn->insertAtTail(new DefExpr(followerIterator));
+
+  if( !zippered ) {
+    fifn->insertAtTail(new CallExpr(PRIM_MOVE, followerIterator, new CallExpr("_toFollower", fifnIterator, fifnFollower)));
+  } else {
+    fifn->insertAtTail(new CallExpr(PRIM_MOVE, followerIterator, new CallExpr("_toFollowerZip", fifnIterator, fifnFollower)));
+  }
+
+  return fifn;
+}
+
+//
+// Lifted from 'createTaskFunctions'
+//
+// TODO: build a more general flattening or outer-finding set of functions
+//
```

Comment:
Please explain the role - and the naming (why "fa"?) of `faParent` in a short comment.

Reply:
Ah, I think that was carried over from when the branch only worked for ForallExprs (ForAll). It could be renamed to something else entirely at this point.